File: tests/cribl_client.py
"""
Cribl management API client for pack testing.

Adapted from the criblpacks/cribl-palo-alto-networks pattern (Cribl's own pack
test pattern). Streamlined for parametrized fixture-based testing.

References:
- https://github.com/criblpacks/cribl-palo-alto-networks/blob/main/test/cribl_stream.py
- https://docs.cribl.io/api-reference/
"""

from __future__ import annotations

import io
import json
import os
import re
import tarfile
import time
import uuid
from json import JSONDecodeError
from pathlib import Path
from typing import Any

import ndjson
import requests
import yaml

# Locate the pack root (one level up from tests/) so client helpers can read
# package.json and route.yml without each caller passing the path in.
PACK_ROOT = Path(__file__).resolve().parent.parent

# Canonical Cribl filter shape we can auto-resolve: `<field>=='<value>'`.
# Anything more complex (boolean ops, function calls, regex) is out of scope
# for the local evaluator — callers should pytest.skip when this returns None.
_SIMPLE_FILTER_RE = re.compile(r"^\s*([A-Za-z_]\w*)\s*==\s*['\"](.*?)['\"]\s*$")


def _parse_simple_filter(expr: str) -> tuple[str, str] | None:
    """Parse `<field>=='<value>'` (or `=="value"`) into (field, value).

    Returns None for any expression we cannot statically resolve. Tests should
    pytest.skip when None is returned so complex filters don't block the suite.
    """
    match = _SIMPLE_FILTER_RE.match(expr or "")
    if not match:
        return None
    return match.group(1), match.group(2)


class CriblClient:
    """Thin wrapper around the Cribl management API for pack testing."""

    def __init__(
        self,
        host: str = "localhost",
        port: int = 9000,
        username: str = "admin",
        password: str = "admin",
        scheme: str = "http",
    ) -> None:
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.scheme = scheme
        self._token: str | None = None

    @property
    def base_url(self) -> str:
        return f"{self.scheme}://{self.host}:{self.port}/api/v1"

    @property
    def token(self) -> str:
        if self._token is None:
            self._token = self._login()
        return self._token

    def _login(self) -> str:
        response = requests.post(
            f"{self.base_url}/auth/login",
            json={"username": self.username, "password": self.password},
            timeout=10,
        )
        response.raise_for_status()
        return response.json()["token"]

    def _call(
        self,
        method: str,
        endpoint: str,
        *,
        pack: str | None = None,
        payload: dict | None = None,
        data: bytes | None = None,
        params: dict | None = None,
        authenticated: bool = True,
    ) -> Any:
        prefix = f"/p/{pack}" if pack else ""
        url = f"{self.base_url}{prefix}{endpoint}"

        headers: dict[str, str] = {}
        if authenticated:
            headers["authorization"] = f"Bearer {self.token}"

        response = requests.request(
            method.upper(),
            url,
            headers=headers,
            params=params,
            data=data,
            json=payload,
            timeout=30,
        )

        try:
            return response.json()
        except JSONDecodeError:
            try:
                return response.json(cls=ndjson.Decoder)
            except Exception:
                return response.content

    # ---- Lifecycle ---------------------------------------------------------

    def wait_until_ready(self, timeout_seconds: int = 120) -> None:
        """Poll the health endpoint until Cribl is ready or we time out."""
        deadline = time.time() + timeout_seconds
        last_err: Exception | None = None
        while time.time() < deadline:
            try:
                response = requests.get(
                    f"{self.base_url}/health",
                    timeout=5,
                )
                if response.status_code == 200:
                    return
            except requests.RequestException as exc:
                last_err = exc
            time.sleep(2)
        raise TimeoutError(
            f"Cribl at {self.base_url} did not become ready in {timeout_seconds}s "
            f"(last error: {last_err})"
        )

    # ---- Pack management ---------------------------------------------------

    @staticmethod
    def create_pack_tarball(pack_root: Path) -> bytes:
        """Build an in-memory .crbl tarball from pack_root.

        Excludes test/dev directories that should not ship inside the pack.
        """
        excluded_basenames = {
            ".git",
            ".github",
            "tests",
            "test",
            "venv",
            ".venv",
            ".DS_Store",
            ".idea",
            ".pytest_cache",
            "__pycache__",
            ".direnv",
        }

        def filter_func(tarinfo: tarfile.TarInfo) -> tarfile.TarInfo | None:
            name = os.path.basename(tarinfo.name)
            if name in excluded_basenames:
                return None
            if name.endswith(".crbl"):
                return None
            tarinfo.uid = tarinfo.gid = 0
            tarinfo.uname = tarinfo.gname = "root"
            return tarinfo

        buffer = io.BytesIO()
        with tarfile.open(fileobj=buffer, mode="w:gz") as tar:
            tar.add(str(pack_root), arcname="", filter=filter_func)
        buffer.seek(0)
        return buffer.read()

    def install_pack(self, tarball: bytes, expected_id: str | None = None, timeout_seconds: int = 30) -> None:
        """Upload a .crbl tarball, install it, and (optionally) wait for it to appear.

        Cribl's pack install is asynchronous — the POST returns before the pack
        is fully registered. If expected_id is given, poll /packs until it
        appears (or raise TimeoutError). Without expected_id, returns
        immediately after the install POST and the caller is responsible for
        waiting if needed.
        """
        params = {
            "filename": f"{uuid.uuid4()}.crbl",
            "size": len(tarball),
        }
        upload = self._call("put", "/packs", params=params, data=tarball)
        if upload:
            self._call("post", "/packs", payload=upload)

        if expected_id is None:
            return

        deadline = time.time() + timeout_seconds
        installed: list[str | None] = []
        while time.time() < deadline:
            installed = [p.get("id") for p in self.list_packs()]
            if expected_id in installed:
                return
            time.sleep(0.5)
        raise TimeoutError(
            f"Pack '{expected_id}' did not appear in /packs after {timeout_seconds}s. "
            f"Currently installed: {installed}"
        )

    def delete_pack(self, pack_id: str) -> None:
        info = self._call("get", f"/packs/{pack_id}")
        if info:
            self._call("delete", f"/packs/{pack_id}", payload=info)

    def list_packs(self) -> list[dict]:
        response = self._call("get", "/packs")
        return response.get("items", []) if isinstance(response, dict) else []

    # ---- Sample lifecycle --------------------------------------------------

    def save_sample(self, name: str, events: list[dict]) -> str:
        """Save events as a named sample, returning the sample ID."""
        payload = {
            "sampleName": name,
            "context": {"events": events},
        }
        response = self._call("post", "/system/samples", payload=payload)
        return response["items"][0]["id"]

    def delete_sample(self, sample_id: str) -> None:
        info = self._call("get", f"/system/samples/{sample_id}")
        if isinstance(info, dict) and info.get("items"):
            self._call(
                "delete",
                f"/system/samples/{sample_id}",
                payload=info["items"][0],
            )

    def delete_all_samples(self) -> None:
        response = self._call("get", "/system/samples")
        if not isinstance(response, dict):
            return
        for sample in response.get("items", []):
            if "isTemplate" in sample:
                continue
            self._call(
                "delete",
                f"/system/samples/{sample['id']}",
                payload=sample,
            )

    # ---- Pipeline execution ------------------------------------------------

    def run_pipeline(
        self,
        pipeline: str,
        sample_id: str,
        pack: str | None = None,
        timeout_ms: int = 10000,
        memory_mb: int = 2048,
    ) -> list[dict]:
        """Execute a pipeline against a saved sample, returning processed events."""
        payload = {
            "mode": "pipe",
            "pipelineId": pipeline,
            "level": 3,
            "sampleId": sample_id,
            "dropped": True,
            "cpuProfile": False,
            "timeout": timeout_ms,
            "memory": memory_mb,
        }
        response = self._call("post", "/preview", pack=pack, payload=payload)
        if isinstance(response, dict):
            return response.get("items", [])
        return response or []

    # ---- Route flow + assertions ------------------------------------------

    def run_route_flow(
        self,
        sample_id: str,
        events: list[dict],
        pack: str | None = None,
    ) -> dict:
        """Match events against route.yml filters, then execute the matched pipeline.

        Cribl's `/preview` API has no `mode: "route"` — this is the local
        fallback. Loads `PACK_ROOT/default/pipelines/route.yml`, evaluates each
        non-disabled route's filter against the supplied events using the
        canonical `<field>=='<value>'` matcher, and on the first match calls
        `run_pipeline()` for that route's pipeline.

        Returns:
            {"route": <route dict>, "pipeline": <pipeline id>, "events": [...]}

        Raises:
            ValueError: if no route matches. The error lists any filters that
                were skipped because the local matcher doesn't support them, so
                the caller can decide whether to treat the situation as a
                pytest.skip vs a real failure.
        """
        route_yml = PACK_ROOT / "default" / "pipelines" / "route.yml"
        config = yaml.safe_load(route_yml.read_text())

        skipped_filters: list[str] = []
        for route in config.get("routes", []):
            if route.get("disabled"):
                continue
            parsed = _parse_simple_filter(route.get("filter", ""))
            if parsed is None:
                skipped_filters.append(route.get("filter", ""))
                continue
            field, expected_value = parsed
            if any(event.get(field) == expected_value for event in events):
                output = self.run_pipeline(route["pipeline"], sample_id, pack=pack)
                return {
                    "route": route,
                    "pipeline": route["pipeline"],
                    "events": output,
                }

        raise ValueError(
            "No matching route for given events. "
            f"Filters skipped (unparseable by local matcher): {skipped_filters}"
        )

    def assert_required_fields(
        self,
        events: list[dict],
        pack_type: str | None = None,
    ) -> None:
        """Assert every event has the canonical output fields for its pack type.

        Edge packs require `sourcetype` + `index` (Splunk-canonical routing
        fields). Stream packs require `host` + `source` + `_time`.

        When `pack_type` is None, infer it from `package.json` name prefix
        (`cc-edge-` -> edge, `cc-stream-` -> stream). Tests can override by
        passing pack_type explicitly.
        """
        if pack_type is None:
            pack_type = _detect_pack_type()
        if pack_type == "edge":
            required: tuple[str, ...] = ("sourcetype", "index")
        elif pack_type == "stream":
            required = ("host", "source", "_time")
        else:
            raise ValueError(
                f"Unknown pack_type {pack_type!r} (expected 'edge' or 'stream')"
            )

        violations: list[str] = []
        for i, event in enumerate(events):
            missing = [k for k in required if k not in event]
            if missing:
                violations.append(f"event {i}: missing {missing}")

        assert not violations, (
            f"{pack_type.title()} pack required fields {required} missing from "
            f"{len(violations)}/{len(events)} event(s):\n  "
            + "\n  ".join(violations)
        )

    # ---- Live capture ------------------------------------------------------
    #
    # Two-phase async API. start_capture() returns a capture id immediately;
    # read_capture() polls until the session reports complete. These primitives
    # exist to support future integration tests that publish into a live source
    # and assert routes fire — Phase A's unit tests don't exercise them.

    def start_capture(
        self,
        filter_expr: str,
        max_events: int = 100,
        timeout_ms: int = 30000,
    ) -> str:
        """Start an async capture matching `filter_expr`. Returns the capture id."""
        payload = {
            "filter": filter_expr,
            "maxEvents": max_events,
            "timeout": timeout_ms,
            "level": 0,
        }
        response = self._call("post", "/lib/captures", payload=payload)
        if not isinstance(response, dict) or "captureId" not in response:
            raise RuntimeError(f"Unexpected capture response: {response!r}")
        return response["captureId"]

    def read_capture(
        self,
        capture_id: str,
        poll_interval: float = 1.0,
        timeout_seconds: int = 60,
    ) -> list[dict]:
        """Poll the capture session until status=='complete', returning events."""
        deadline = time.time() + timeout_seconds
        while time.time() < deadline:
            response = self._call("get", f"/lib/captures/{capture_id}/events")
            if isinstance(response, dict) and response.get("status") == "complete":
                return response.get("items", [])
            time.sleep(poll_interval)
        raise TimeoutError(
            f"Capture {capture_id} did not complete in {timeout_seconds}s"
        )


def _detect_pack_type() -> str:
    """Infer pack_type ('edge' or 'stream') from package.json name."""
    name = json.loads((PACK_ROOT / "package.json").read_text()).get("name", "")
    if name.startswith("cc-edge-"):
        return "edge"
    if name.startswith("cc-stream-"):
        return "stream"
    raise ValueError(
        f"Cannot detect pack_type from package.json name {name!r}; "
        "expected 'cc-edge-<source>-io' or 'cc-stream-<source>-io' prefix."
    )

File: tests/cribl-client.ts
/**
 * Cribl management API client for pack testing.
 *
 * TypeScript port of the criblpacks/cribl-palo-alto-networks pattern.
 * Streamlined for Vitest-based fixture testing.
 *
 * References:
 * - https://docs.cribl.io/api-reference/
 * - https://github.com/criblpacks/cribl-palo-alto-networks
 */

import { randomUUID } from "node:crypto";
import { readFileSync } from "node:fs";
import { readdir, readFile } from "node:fs/promises";
import * as path from "node:path";
import { setTimeout as sleep } from "node:timers/promises";
import { create as tarCreate } from "tar";
import { parse as yamlParse } from "yaml";
import { parseSimpleFilter } from "./parse-filter.js";

// PACK_ROOT is the directory containing default/, data/, package.json — i.e.
// one level up from this tests/ folder.
export const PACK_ROOT = path.resolve(import.meta.dirname, "..");

export interface CriblClientOptions {
  host?: string | undefined;
  port?: number | undefined;
  username?: string | undefined;
  password?: string | undefined;
  scheme?: "http" | "https" | undefined;
}

export interface CriblEvent {
  [key: string]: unknown;
}

export interface Route {
  id?: string;
  name?: string;
  pipeline: string;
  filter?: string;
  output?: string;
  disabled?: boolean;
  [key: string]: unknown;
}

export interface RouteFlowResult {
  route: Route;
  pipeline: string;
  events: CriblEvent[];
}

export type PackType = "edge" | "stream";

const REQUIRED_FIELDS: Record<PackType, readonly string[]> = {
  edge: ["sourcetype", "index"],
  stream: ["host", "source", "_time"],
};

// Cribl pack contents — only these top-level entries ship inside the .crbl.
// Whitelisting (rather than blacklisting tooling files) keeps the tarball
// stable as we add repo-level dev tooling (flake.nix, biome.jsonc, etc.) that
// shouldn't leak into the pack distribution. Passed as directory names to
// node-tar's `create()`, which auto-recurses + emits proper directory
// entries (Cribl rejects tarballs missing them).
const PACK_ROOT_ENTRIES = new Set([
  "data",
  "default",
  "package.json",
  "README.md",
  "LICENSE",
]);

export class CriblClient {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly scheme: string;
  private cachedToken: string | null = null;

  constructor(options: CriblClientOptions = {}) {
    this.host = options.host ?? "localhost";
    this.port = options.port ?? 9000;
    this.username = options.username ?? "admin";
    this.password = options.password ?? "admin";
    this.scheme = options.scheme ?? "http";
  }

  get baseUrl(): string {
    return `${this.scheme}://${this.host}:${this.port}/api/v1`;
  }

  private async login(): Promise<string> {
    const response = await fetch(`${this.baseUrl}/auth/login`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        username: this.username,
        password: this.password,
      }),
    });
    if (!response.ok) {
      throw new Error(
        `Cribl login failed: HTTP ${response.status} ${await response.text()}`,
      );
    }
    const body = (await response.json()) as { token: string };
    return body.token;
  }

  async token(): Promise<string> {
    if (this.cachedToken === null) {
      this.cachedToken = await this.login();
    }
    return this.cachedToken;
  }

  /**
   * Send an authenticated HTTP request to the Cribl management API.
   *
   * The `pack` option scopes the URL to `/p/<pack>/<endpoint>` (matches
   * Cribl's pack-namespaced preview routes).
   */
  async call(
    method: string,
    endpoint: string,
    options: {
      pack?: string | undefined;
      payload?: unknown;
      body?: string | Uint8Array | undefined;
      params?: Record<string, string | number> | undefined;
      authenticated?: boolean | undefined;
      contentType?: string | undefined;
    } = {},
  ): Promise<unknown> {
    const prefix = options.pack !== undefined ? `/p/${options.pack}` : "";
    const search =
      options.params !== undefined
        ? `?${new URLSearchParams(
            Object.fromEntries(
              Object.entries(options.params).map(([k, v]) => [k, String(v)]),
            ),
          ).toString()}`
        : "";
    const url = `${this.baseUrl}${prefix}${endpoint}${search}`;

    const headers: Record<string, string> = {};
    if (options.authenticated !== false) {
      headers.authorization = `Bearer ${await this.token()}`;
    }

    let body: string | Uint8Array | undefined;
    if (options.payload !== undefined) {
      headers["content-type"] = options.contentType ?? "application/json";
      body = JSON.stringify(options.payload);
    } else if (options.body !== undefined) {
      headers["content-type"] =
        options.contentType ?? "application/octet-stream";
      body = options.body;
    }

    const init: RequestInit = { method: method.toUpperCase(), headers };
    if (body !== undefined) init.body = body;
    const response = await fetch(url, init);
    const text = await response.text();

    if (!response.ok) {
      throw new Error(
        `Cribl API ${method.toUpperCase()} ${endpoint} failed: HTTP ${response.status} ${response.statusText} — ${text || "<empty body>"}`,
      );
    }

    // Try JSON first; fall back to NDJSON; fall back to raw text.
    if (text.length === 0) return null;
    try {
      return JSON.parse(text);
    } catch {
      // NDJSON (newline-delimited)
      const lines = text.split(/\r?\n/).filter((l) => l.length > 0);
      try {
        return lines.map((line) => JSON.parse(line));
      } catch {
        return text;
      }
    }
  }

  // ---- Lifecycle -------------------------------------------------------

  async waitUntilReady(timeoutSeconds = 120): Promise<void> {
    const deadline = Date.now() + timeoutSeconds * 1000;
    let lastError: unknown;
    while (Date.now() < deadline) {
      try {
        const response = await fetch(`${this.baseUrl}/health`);
        if (response.status === 200) return;
      } catch (err) {
        lastError = err;
      }
      await sleep(2000);
    }
    throw new Error(
      `Cribl at ${this.baseUrl} did not become ready in ${timeoutSeconds}s (last error: ${String(lastError)})`,
    );
  }

  // ---- Pack management -------------------------------------------------

  /**
   * Build an in-memory .crbl tarball from packRoot, excluding test/dev directories.
   */
  static async createPackTarball(packRoot: string): Promise<Buffer> {
    const entries = await collectTarballEntries(packRoot);
    const stream = tarCreate(
      {
        gzip: true,
        cwd: packRoot,
        portable: true,
        // Normalize ownership so the tarball is deterministic.
        mtime: new Date(0),
      },
      entries,
    );
    const chunks: Buffer[] = [];
    for await (const chunk of stream) {
      chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
    }
    return Buffer.concat(chunks);
  }

  /**
   * Upload a .crbl tarball, install it, and (optionally) wait for it to register.
   *
   * Cribl's pack install is asynchronous — the POST returns before the pack
   * is fully registered. If `expectedId` is provided, poll /packs until it
   * appears (or throw on timeout). Without `expectedId`, returns immediately
   * after the install POST.
   */
  async installPack(
    tarball: Buffer,
    expectedId?: string,
    timeoutSeconds = 30,
  ): Promise<void> {
    const filename = `${randomUUID()}.crbl`;
    const upload = await this.call("put", "/packs", {
      params: { filename, size: tarball.length },
      body: new Uint8Array(tarball),
    });
    if (upload !== null && upload !== undefined) {
      await this.call("post", "/packs", { payload: upload });
    }

    if (expectedId === undefined) return;

    const deadline = Date.now() + timeoutSeconds * 1000;
    let installed: (string | undefined)[] = [];
    while (Date.now() < deadline) {
      installed = (await this.listPacks()).map((p) => p.id);
      if (installed.includes(expectedId)) return;
      await sleep(500);
    }
    throw new Error(
      `Pack '${expectedId}' did not appear in /packs after ${timeoutSeconds}s. Currently installed: ${JSON.stringify(installed)}`,
    );
  }

  async deletePack(packId: string): Promise<void> {
    const info = await this.call("get", `/packs/${packId}`);
    if (info !== null && info !== undefined) {
      await this.call("delete", `/packs/${packId}`, { payload: info });
    }
  }

  async listPacks(): Promise<{ id?: string }[]> {
    const response = await this.call("get", "/packs");
    if (
      response !== null &&
      typeof response === "object" &&
      "items" in response
    ) {
      const items = (response as { items: unknown }).items;
      if (Array.isArray(items)) return items as { id?: string }[];
    }
    return [];
  }

  // ---- Sample lifecycle -----------------------------------------------

  async saveSample(name: string, events: CriblEvent[]): Promise<string> {
    const response = (await this.call("post", "/system/samples", {
      payload: { sampleName: name, context: { events } },
    })) as { items: { id: string }[] };
    if (response.items[0] === undefined) {
      throw new Error(`saveSample('${name}') returned no items`);
    }
    return response.items[0].id;
  }

  async deleteSample(sampleId: string): Promise<void> {
    const info = await this.call("get", `/system/samples/${sampleId}`);
    if (
      info !== null &&
      typeof info === "object" &&
      "items" in info &&
      Array.isArray((info as { items: unknown }).items) &&
      (info as { items: unknown[] }).items.length > 0
    ) {
      await this.call("delete", `/system/samples/${sampleId}`, {
        payload: (info as { items: unknown[] }).items[0],
      });
    }
  }

  async deleteAllSamples(): Promise<void> {
    const response = await this.call("get", "/system/samples");
    if (
      response === null ||
      typeof response !== "object" ||
      !("items" in response)
    )
      return;
    const items = (response as { items: unknown[] }).items;
    if (!Array.isArray(items)) return;
    for (const sample of items) {
      if (
        sample !== null &&
        typeof sample === "object" &&
        "isTemplate" in sample
      )
        continue;
      const id = (sample as { id?: string }).id;
      if (id !== undefined) {
        await this.call("delete", `/system/samples/${id}`, { payload: sample });
      }
    }
  }

  // ---- Pipeline execution ---------------------------------------------

  async runPipeline(
    pipeline: string,
    sampleId: string,
    options: {
      pack?: string | undefined;
      timeoutMs?: number;
      memoryMb?: number;
    } = {},
  ): Promise<CriblEvent[]> {
    const response = await this.call("post", "/preview", {
      pack: options.pack,
      payload: {
        mode: "pipe",
        pipelineId: pipeline,
        level: 3,
        sampleId,
        dropped: true,
        cpuProfile: false,
        timeout: options.timeoutMs ?? 10_000,
        memory: options.memoryMb ?? 2048,
      },
    });
    if (
      response !== null &&
      typeof response === "object" &&
      "items" in response
    ) {
      const items = (response as { items: unknown }).items;
      if (Array.isArray(items)) return items as CriblEvent[];
    }
    if (Array.isArray(response)) return response as CriblEvent[];
    return [];
  }

  // ---- Route flow + assertions ----------------------------------------

  /**
   * Match events against route.yml filters, then execute the matched pipeline.
   *
   * Cribl's `/preview` API has no `mode: "route"` — this is the local
   * fallback. Loads `PACK_ROOT/default/pipelines/route.yml`, evaluates each
   * non-disabled route's filter via `parseSimpleFilter` (canonical
   * `<field>=='<value>'` matcher), and on the first match calls
   * `runPipeline` for that route's pipeline.
   *
   * Throws if no route matches (lists any filters skipped because the local
   * matcher doesn't support them, so the caller can decide whether to skip
   * the test vs treat it as a real failure).
   */
  async runRouteFlow(
    sampleId: string,
    events: CriblEvent[],
    options: { pack?: string | undefined } = {},
  ): Promise<RouteFlowResult> {
    const routeYml = path.join(PACK_ROOT, "default", "pipelines", "route.yml");
    const config = yamlParse(await readFile(routeYml, "utf-8")) as {
      routes?: Route[];
    };

    const skippedFilters: string[] = [];
    for (const route of config.routes ?? []) {
      if (route.disabled === true) continue;
      const parsed = parseSimpleFilter(route.filter);
      if (parsed === null) {
        skippedFilters.push(route.filter ?? "");
        continue;
      }
      const [field, expectedValue] = parsed;
      const matches = events.some((event) => event[field] === expectedValue);
      if (matches) {
        const output = await this.runPipeline(
          route.pipeline,
          sampleId,
          options,
        );
        return { route, pipeline: route.pipeline, events: output };
      }
    }

    throw new Error(
      `No matching route for given events. Filters skipped (unparseable by local matcher): ${JSON.stringify(skippedFilters)}`,
    );
  }

  /**
   * Assert every event has the canonical output fields for its pack type.
   *
   * Edge packs require `sourcetype` + `index` (Splunk-canonical routing fields).
   * Stream packs require `host` + `source` + `_time`.
   *
   * When `packType` is undefined, infer from `package.json` name prefix.
   */
  static assertRequiredFields(events: CriblEvent[], packType?: PackType): void {
    const resolvedType = packType ?? detectPackType();
    const required = REQUIRED_FIELDS[resolvedType];

    const violations: string[] = [];
    for (const [i, event] of events.entries()) {
      const missing = required.filter((field) => !(field in event));
      if (missing.length > 0) {
        violations.push(`event ${i}: missing ${JSON.stringify(missing)}`);
      }
    }

    if (violations.length > 0) {
      throw new Error(
        `${resolvedType} pack required fields ${JSON.stringify(required)} missing from ${violations.length}/${events.length} event(s):\n  ${violations.join("\n  ")}`,
      );
    }
  }

  // ---- Live capture (primitives reserved for future integration tests) -

  async startCapture(
    filterExpr: string,
    maxEvents = 100,
    timeoutMs = 30_000,
  ): Promise<string> {
    const response = await this.call("post", "/lib/captures", {
      payload: { filter: filterExpr, maxEvents, timeout: timeoutMs, level: 0 },
    });
    if (
      response === null ||
      typeof response !== "object" ||
      !("captureId" in response)
    ) {
      throw new Error(
        `Unexpected capture response: ${JSON.stringify(response)}`,
      );
    }
    return (response as { captureId: string }).captureId;
  }

  async readCapture(
    captureId: string,
    options: { pollIntervalMs?: number; timeoutSeconds?: number } = {},
  ): Promise<CriblEvent[]> {
    const pollIntervalMs = options.pollIntervalMs ?? 1000;
    const timeoutSeconds = options.timeoutSeconds ?? 60;
    const deadline = Date.now() + timeoutSeconds * 1000;
    while (Date.now() < deadline) {
      const response = await this.call(
        "get",
        `/lib/captures/${captureId}/events`,
      );
      if (
        response !== null &&
        typeof response === "object" &&
        "status" in response &&
        (response as { status: string }).status === "complete"
      ) {
        const items = (response as { items?: unknown }).items;
        return Array.isArray(items) ? (items as CriblEvent[]) : [];
      }
      await sleep(pollIntervalMs);
    }
    throw new Error(
      `Capture ${captureId} did not complete in ${timeoutSeconds}s`,
    );
  }
}

/**
 * Infer pack type from the pack root's package.json name field.
 */
export function detectPackType(): PackType {
  const pkg = JSON.parse(
    readFileSync(path.join(PACK_ROOT, "package.json"), "utf-8"),
  ) as {
    name?: string;
  };
  const name = pkg.name ?? "";
  if (name.startsWith("cc-edge-")) return "edge";
  if (name.startsWith("cc-stream-")) return "stream";
  throw new Error(
    `Cannot detect pack_type from package.json name '${name}'; expected 'cc-edge-<source>-io' or 'cc-stream-<source>-io' prefix.`,
  );
}

/**
 * Read the pack id (package.json name) — used by tests + setup.
 */
export function getPackId(): string {
  const pkg = JSON.parse(
    readFileSync(path.join(PACK_ROOT, "package.json"), "utf-8"),
  ) as {
    name?: string;
  };
  if (pkg.name === undefined || pkg.name.length === 0) {
    throw new Error('package.json missing required "name" field');
  }
  return pkg.name;
}

// ---- Internal helpers --------------------------------------------------

async function collectTarballEntries(packRoot: string): Promise<string[]> {
  const dirents = await readdir(packRoot, { withFileTypes: true });
  return dirents
    .filter((d) => PACK_ROOT_ENTRIES.has(d.name))
    .map((d) => d.name);
}
